node/lib/util/reset.js

```diff
@@ -69,49 +75,69 @@ function getType(type) {
 
         case TYPE.MERGE: return NodeGit.Reset.TYPE.HARD;
     }
-    assert(false, `Bad type: ${type}`);
 }
 
 /**
- * Throw a `UserError` if any open submodule as listed by the specified
- * `opener` has unstaged changes and does not exist in the specified `shas`.
+ * Reset the specified `repo` of having  specified `index` to have the contents
+ * of the tree of the specified `commit`.  Update the `.gitmodules` file in the
+ * worktree to haave the same contents as in the index.  If the repo is not in
+ * sparse mode, create empty directories for added submodules and remove the
+ * directories of deleted submodules as indicated in the specified `changes`.
  *
- * @param {Open.Opener} opener
- * @param {Object}      shas      map from submodule path to SHA
+ * @param {NodeGit.Repository} repo
+ * @param {NodeGit.Index}      index
+ * @param {NodeGit.Commit}     commit
+ * @param {Object}             changes     from path to `SubmoduleChange`
  */
-exports.validateClean = co.wrap(function *(opener, shas) {
-    assert.instanceOf(opener, Open.Opener);
-    assert.isObject(shas);
-
-    const openSubs = Array.from(yield opener.getOpenSubs());
-    const badSubs = [];
-    const checkSub = co.wrap(function *(name) {
-        if (name in shas) {
-            // If it exists in `shas` we don't need to check it
-            return;                                                   // RETURN
+exports.resetMetaRepo = co.wrap(function *(repo, index, commit, changes) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(index, NodeGit.Index);
+    assert.instanceOf(commit, NodeGit.Commit);
+    assert.isObject(changes);
+
+    const tree = yield commit.getTree();
+    yield index.readTree(tree);
+
+    // Render modules file
+
+    const modulesFileName = SubmoduleConfigUtil.modulesFileName;
+    const modulesPath = path.join(repo.workdir(), modulesFileName);
+    const modulesEntry = index.getByPath(modulesFileName);
+    if (undefined !== modulesEntry) {
+        const oid = modulesEntry.id;
+        const blob = yield repo.getBlob(oid);
+        const data = blob.toString();
+        yield fs.writeFile(modulesPath, data);
+    } else {
+        // If it's not in the index, remove it.
+        try {
+            yield fs.unlink(modulesPath);
+        } catch (e) {
+            if ("ENOENT" !== e.code) {
+                throw e;
+            }
         }
-        const repo = yield opener.getSubrepo(name);
-        const status = yield StatusUtil.getRepoStatus(repo, {
-            showMetaChanges: true,
+    }
+
+    // Tidy up directories when not in sparse mode.
+
+    if (!(yield SparseCheckoutUtil.inSparseMode(repo))) {
```

Comment:
If we're in sparse mode, why don't we do this?

Reply:
Well spell out reason in comment.